New CLI

gicaceres opened the pull request:
Commands

- [ x ] - a8ctl-beta service-list
- [ x ] - a8ctl-beta route-list
- [ x ] - a8ctl-beta action-list
- [ x ] - a8ctl-beta rule-create
- [ x ] - a8ctl-beta rule-update
- [ x ] - a8ctl-beta rule-get
- [ x ] - a8ctl-beta rule-delete
- [ x ] - a8ctl-beta traffic-start
- [ x ] - a8ctl-beta traffic-step
- [ x ] - a8ctl-beta traffic-abort
- [ x ] - a8ctl-beta recipe-run
- [ x ] - a8ctl-beta info

I included the following files to help users with the transition from the old CLI while we update the official documentation.

README.md
HELLOWORLD.md
BOOKINFO.md

gicaceres committed:
Added new a8cli

gicaceres committed:
Updates made by goimports

gicaceres committed:
Updates to make the cli part of the main a8 repo

gicaceres committed:
Changed action-list to display the same output as the current a8ctl

gicaceres committed:
Added traffic-start

gicaceres committed:
Added traffic-step

gicaceres committed:
Added traffic-abort

gicaceres committed:
rule-get returns yaml by default

This closes #410

gicaceres committed:
Added recipe-run

This closes #414

gicaceres committed:
Re-word message

gicaceres committed:
Added support to terminate input by entering EOF in rule-create.
(CTRL+D) on Unix-like systems or (Ctrl+Z) in Windows.

This closes #412

gicaceres committed:
Added rule-update

gicaceres committed:
Update README file

gicaceres committed:
Support input redirection to create and update rules

This closes #411

gicaceres committed:
Filename typo

gicaceres committed:
CLI command equivalents for HelloWorld and Bookinfo examples

rshriram commented on Makefile:
```diff
@@ -53,13 +54,14 @@ SIDECAR_DOCKERFILE		:= $(DOCKERDIR)/Dockerfile.sidecar.ubuntu
 CONTROLLER_RELEASE_NAME	:= $(CONTROLLER_APP_NAME)-$(APP_VER)-$(GOOS)-$(GOARCH)
 SIDECAR_RELEASE_NAME	:= $(SIDECAR_APP_NAME)-$(APP_VER)-$(GOOS)-$(GOARCH)
+CLI_RELEASE_NAME	:= $(CLI_APP_NAME)-$(APP_VER)-$(GOOS)-$(GOARCH)
 
```
The entire purpose of the CLI was to achieve platform independency and make it easier for people to just download and use a single binary. So, in that spirit, we need 3 binaries
a8ctl-linux-amd64
a8ctl-osx-amd64
a8ctl-windows-amd64

gicaceres replied:
done

rshriram commented on Makefile:
```diff
@@ -107,6 +109,12 @@ build.sidecar:
+	@go-bindata -pkg=utils -prefix "./cli" -o ./cli/utils/i18n_resources.go ./cli/locales
+	@go build $(BUILDFLAGS) -ldflags '$(LDFLAGS)' -o $(BINDIR)/$(CLI_APP_NAME) ./cmd/cli/
+	@goimports -w ./cli/utils/i18n_resources.go
+
```
See comment above. Need to cross compile for 3 platforms.

gicaceres replied:
done

rshriram commented on Makefile:
```diff
@@ -194,37 +202,40 @@ dockerize.sidecar:
+
+compress.cli: tools.upx
+	@make --no-print-directory compress COMPRESSED_FILE=$(BINDIR)/$(CLI_APP_NAME)
+
```
No need to compress CLI.

gicaceres replied:
done

rshriram commented on Makefile:
```diff
@@ -296,3 +312,9 @@ tools.upx:
 	fi
+
+tools.go-bindata:
+	@command -v go-bindata >/dev/null ; if [ $$? -ne 0 ]; then \
```
some comments here would be useful on what go-bindata does

gicaceres replied:
done

rshriram commented on cli/api/controller_client.go:
```diff
@@ -0,0 +1,166 @@
+		client: api.NewClient(url, token, client),
+	}, nil
+}
+
```
Is this using the newly refactored controller client?

gicaceres replied:
No, it's not using the newly refactored controller client.

When I started working on the CLI, I wrote the controller and registry clients based on the idea that the new CLI was going to have its own repo, like the python version.

Now that I have updated the code to make the CLI part of the main A8 repo and added more commands, it makes more sense to reuse/update code from/to other packages so I will start looking at those changes.

Since we are releasing the CLI as beta version, I think is okay to merge the CLI as it is for now. We can remove the BETA tag once the CLI is more integrated with the existing code along with the corresponding doc updates and IT.

elevran replied:
Please see general comment below regarding "Beta" designation.

rshriram commented on cli/api/controller_structs.go:
```diff
@@ -0,0 +1,99 @@
+	Destination string        `json:"destination"`
+	Match       *MatchRules   `json:"match"`
+	Actions     *ActionsRules `json:"actions"`
+}
```
Doesn't the controller already handle this today? If not, this should actually go into pkg/api/rules

gicaceres replied:
See comment above.

rshriram commented on cli/api/registry_client.go:
```diff
@@ -0,0 +1,78 @@
+		return url, common.ErrRegistryURLInvalid
+	}
+	return url, nil
+}
```
can reuse pkg/discovery I think. See code in sidecar for examples

gicaceres replied:
See comment above.

rshriram reviewed (changes requested):
Overall this is a good first pass. There is lots of code that can either be reused from existing packages or be moved into common shared packages (c.f. comments related to registry/controller rules structs). 

We also need cross compilation as part of Makefile.

rshriram commented:
I don't think we should be merging code that is using custom clients to
controller and registry. There has been and will continue to be many
changes to the two components. The cli may break at any time.

Secondly as part of this series, we need to add this cli to the integration
tests.

@zcahana ^^^

On Fri, Dec 9, 2016 at 5:34 PM Gerson Caceres <notifications@github.com>
wrote:

> *@gicaceres* commented on this pull request.
> ------------------------------
>
> In cli/api/registry_client.go
> <https://github.com/amalgam8/amalgam8/pull/433>:
>
> > +		return nil, err
> +	}
> +	return instances, nil
> +}
> +
> +// ValidateRegistryURL .
> +func ValidateRegistryURL(ctx *cli.Context) (string, error) {
> +	url := ctx.GlobalString(common.RegistryURL.Flag())
> +	if len(url) == 0 {
> +		return "empty", common.ErrRegistryURLNotFound
> +	}
> +	if !utils.IsURL(url) {
> +		return url, common.ErrRegistryURLInvalid
> +	}
> +	return url, nil
> +}
>
> See comment above.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/amalgam8/amalgam8/pull/433>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH0qd6zAVS5xmrUakLCA0CIe76xtJLFpks5rGdd2gaJpZM4LHIwA>
> .
>
-- 

~shriram

gicaceres committed:
Switch to A8 Registry client

GregHanson commented on registry/client/client.go:
```diff
@@ -261,3 +265,26 @@ func normalizeConfig(config *Config) error {
+}
+
+// printCurl prints the curl of a given request
+func (client *Client) printCurl(method, url string, data interface{}, headers http.Header) {
```
Pretty sure there is at least one other printCurl() function declaration so perhaps this should be moved to a util function

gicaceres replied:
At this time, there is 1 declaration in the A8 registry client and another one inside of the CLI folder for custom ctrl client, until we can switch to the A8 ctrl client. Where do you suggest that we can move this function so it can be use by both clients?

gicaceres replied:
I can create a new file in `pkg/http_client/httpClient.go` and include:
```
- New
- setAuthHeader
- normalizeConfig
- doRequest
- printCurl
- etc

In the CLI folder, I have an http_client shared by the registry, controller and gremlin clients. We should have something like that.

What do you think?

However, this should be done and merged along with the ctrl client.

elevran commented:
General comment - note sure what releasing in Beta implies in reality.

Suggestion: allow merge into main BUT exclude building the new CLI automatically from Makefile.

That way, users may build it explicitly if needed (e.g., testing) and are well aware that it is not released officially.
It can be added into a release once we have sufficient documentation and integration testing for the new CLI.
Following that (or a similar scheme) would also allow smaller PR's to master as code is ready without impacting users. That way, the individual commits (which seem to be concise and well structured, BTW) would have been easier to review.
Thoughts?

elevran commented on cli/a8cli.go:
```diff
@@ -0,0 +1,128 @@
+
+func init() {
+	// Set log level
+	logrus.SetLevel(logrus.ErrorLevel)
```
What is the benefit of having an init() func? Can't the same be done in the beginning of main()?

gicaceres replied:
No difference. I included in the main()

elevran commented on cli/a8cli.go:
```diff
@@ -0,0 +1,128 @@
+	if err != nil {
+		logrus.WithError(err).Debug("Error")
+
+		if strings.Contains(err.Error(), common.ErrUnknowFlag.Error()) {
```
Relying on String contents may be brittle. Consider using and asserting error types instead.

elevran commented on cli/api/controller_client.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+// ControllerClient .
+type ControllerClient interface {
```
Seems like each method (or group) would be used by different code path. It would be easier (e.g., for mock and testing) to separate into different interfaces and combine by embedding.

gicaceres replied:
This custom client is going to be deleted once the official a8 ctrl client provides more functionality.

I don't think we need to make changes to this and related files (e.g., client/client.go) because we are not going to include them in the final version of the CLI.

elevran commented on cli/api/controller_client.go:
```diff
@@ -0,0 +1,116 @@
+	"net/http"
+)
+
+type controller struct {
```
Controller may be overloaded. RoutingRulesController?
AFAIU, the counterpart is called registry_client?

elevran commented on cli/client/client.go:
```diff
@@ -0,0 +1,263 @@
+}
+
+// A8client .
+type A8client struct {
```
No need to export A8Client, can be a8client

elevran commented on cli/client/client.go:
```diff
@@ -0,0 +1,263 @@
+}
+
+// SetRegistryURL .
+func (c *A8client) SetRegistryURL(baseURL string) {
```
If there is no reason to switch this at runtime, prefer initialization on New only and remove these. If keeping, should test for empty strings.

gicaceres replied:
See my previous comment.

elevran commented on cli/client/client.go:
```diff
@@ -0,0 +1,263 @@
+}
+
+// SetHTTPClient .
+func (c *A8client) SetHTTPClient(client *http.Client) {
```
Ditto - prefer in New only.
Fail this on nil client?
Other methods don't check for nil.

gicaceres replied:
See my previous comment.

rshriram replied:
I think the comment here has nothing to do with code duplication. @elevran is talking about handling error conditions (whether or not there is code duplication).

elevran commented on cli/client/client.go:
```diff
@@ -0,0 +1,263 @@
+	// If not http client has been defined, use the default HTTP client
+	if HTTPclient == nil {
+		HTTPclient = &http.Client{
+			Timeout: 30 * time.Second,
```
TLS config needed or default ok?

gicaceres replied:
See my previous comment.

elevran commented on cli/client/client.go:
```diff
@@ -0,0 +1,263 @@
+)
+
+// Client .
+type Client interface {
```
Won't the default http client do? If not, consider making this interface compatible with std client (e.g., Get, Post, etc.)
It may be sufficient to use the client's Do method for all of these and define a Doer interface with Do(req) (resp, err) only.

gicaceres replied:
See my previous comment.

elevran commented on cli/commands/route_list.go:
```diff
@@ -0,0 +1,255 @@
+func (cmd *RouteListCommand) GetMetadata() cli.Command {
+	T := utils.Language(common.DefaultLanguage)
+	return cli.Command{
+		Name:        T("route_list_name"),
```
IIRC, T() allows internationalization. I don't expect command names to be localized.
This applies to all commands...

gicaceres replied:
done

elevran commented on cli/commands/common.go:
```diff
@@ -0,0 +1,55 @@
+	TABLE = "table"
+)
+
+// CommandTable .
```
https://github.com/olekukonko/tablewriter may be an easier alternative than pretty printing code in commands

gicaceres replied:
Our implementation and tablewriter seems to have the approach. The both use 2 parms.
 
header: []string{}
body: [][]string{}

I don't think we need to add a new library for now, but we could switch to this or any other library in the future.

elevran commented on cli/terminal/color.go:
```diff
@@ -0,0 +1,57 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package terminal
```
There are existing terminal packages with support for ANSI colors, etc.
Were they considered?

gicaceres replied:
We don't need to support colors anymore, or we can add this later in another PR. I removed this file to avoid confusion.

elevran commented on cli/terminal/table.go:
```diff
@@ -0,0 +1,80 @@
+// | cell 11   | cell 12   | cell 13   |
+// | cell 21   | cell 22   | cell 23   |
+// +-----------+-----------+-----------+.
+func (t *term) PrintTable(header []string, body [][]string) {
```
See previous comments regarding existing terminal UI and text table formatter packages.

gicaceres replied:
See my previous comments.

elevran commented on cli/utils/i18n.go:
```diff
@@ -0,0 +1,75 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package utils
```
Can i18n be reused from existing sources?
I guess this may have been done pre the monorepo and wanted to avoid cross repo dependencies.

gicaceres replied:
The CLI compiles the locales before building the binary which allows copying/moving the binary to any folder. Since the existing code does not provide this functionality, it can not be reused.

If we really want to add this functionality for other packages, we should do it later in a different PR.

elevran commented on cli/utils/url.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+// IsURL returns true if the given string is a URL
+func IsURL(rawURL string) bool {
```
Consider dropping this. Wrapping ParseRequestURI (or just url.Parse) when it is used only twice seems like overhead to me (considering that it comes with its own test file and used only twice in non test code).

gicaceres replied:
done

elevran reviewed (changes requested):
Please review comments and address. 
Specifically, I'd prefer it not be merged until changed to not build by default (see general comment regarding beta status and implications)

gicaceres committed:
Remove ParseRequestURI wrapper

gicaceres committed:
Do not translate command names

gicaceres committed:
Added CLI to Makefile. Generate binaries for linux, windows and darwin

gicaceres commented:
In this case, we are calling it Beta because it still has a custom controller client, we need integration tests and official documentation.

For the integration test, I updated the demo script `testing/test-scripts/demo_script.sh` to verify that the new CLI works as expected and can replace the old one.

For the documentation, I wrote 3 documents (`README.md`, `HELLOWORLD.md`, `BOOKINFO.md` which are in the `cli` folder) that describe all the commands and how to use them with the HelloWorld and BookInfo examples. Those docs are available for reference only but will help us to  update the official documentation.

I think it is a good idea to remove the new CLI from the Makefile. I reordered the commits so we can easily revert the last 2, just in case.

rshriram commented:
Have you tested the windows version of the binary against minikube or docker toolbox ?

rshriram commented:
See my comments in the Makefile please. GitHub is not displaying them properly inline.

gicaceres commented:
@rshriram 

- Removed the CLI from universal release target.
- The ldflags seems to work for Windows builds. (Tested on Windows 7 Pro.)
- Users need to build the CLI explicitly for the moment. They can still use `make build.cli`
- Removed new CLI from Integration Test script

rshriram commented:
fixes #381
